### PR TITLE
Refactor DAB-DETR in MMDetection 3.x

### PR DESCRIPTION
**Pull request by KeiChiTse**
## Motivation

This PR **resolves** several _confilcits_ and **rearrange** the code structure based on PR #8811. Comparing to PR #8533 for 2.x, this PR aims to refactor DAB-DETR in MMDetection 3.x.

I'm working together with @jshilong, @Li-Qingyun, and @LYMDLUT to refactor the DETR-like algorithms in MMDetection 3.x, using a more lightweight and easy-to-read code style.

Details can be refered to in brother PRs #9149, #8362, #8763, #8544, #9248, feel free to contact me and give your advice!

### CODE REVIEW

**Comment by JosonChan1998** [comment]
Hi !@KeiChiTse, any plans about Anchor DETR?

**Comment by KeiChiTse** [comment]
### [Update]
Fixed error after refactoring and finished aligning loss and inference results compared to 2.x version! The code is ready for training.
**More refactored detr-like algorithms are on the way!**

**Comment by KeiChiTse** [comment]
> > > Hi !@KeiChiTse, any plans about Anchor DETR?
> > 
> > 
> > Hi，thank you for your attention，we will fisrt support Condational DETR、DAB DETR and DINO，then Group DETR and deformable dab is at second position. Now we have met some difficulty to align with original repo, we will try our best.As Row-Column Attention is not commonly used,we may not officially support it at very soon.
> 
> Thanks for your reply!

Guess my partner had done his job haha. Thank you anyway!

File: mmdet/models/detectors/dab_detr.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    def __init__(self,
+                 *args,
+                 iter_update: bool = True,
+                 random_refpoints_xy: bool = False,
+                 num_patterns: int = 0,
+                 **kwargs) -> None:
+        self.iter_update = iter_update
+        self.random_refpoints_xy = random_refpoints_xy
+        self.num_patterns = num_patterns
+        if not isinstance(num_patterns, int):
```
Comment (jshilong) [inline mmdet/models/detectors/dab_detr.py:43]: assert isinstance(num_patterns, int), f'num_patterns should be int but {num_parrerns}'.

mmdetection recommends using f-string.

File: mmdet/models/detectors/dab_detr.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            Warning('num_patterns should be int but {}'.format(
+                type(num_patterns)))
+            self.num_patterns = 0
+
+        super().__init__(*args, **kwargs)
+
+    def _init_layers(self) -> None:
+        """Initialize layers except for backbone, neck and bbox_head."""
+        self.positional_encoding = SinePositionalEncodingHW(
```
Comment (jshilong) [inline mmdet/models/detectors/dab_detr.py:52]: Maybe we can use our own `SinPositionEncoding`?

File: mmdet/models/detectors/dab_detr.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                inverse_sigmoid(self.query_embedding.weight.data[:, :2])
+            self.query_embedding.weight.data[:, :2].requires_grad = False
+
+    def pre_decoder(self, memory: Tensor) -> Tuple[Dict, Dict]:
+        """Prepare intermediate variables before entering Transformer decoder,
+        such as `query`, `query_pos`.
+
+        Args:
+            memory (Tensor): The output embeddings of the Transformer encoder,
+                has shape (num_feat, bs, dim).
```
Comment (jshilong) [inline mmdet/models/detectors/dab_detr.py:83]: num_feat_points

File: mmdet/models/detectors/dab_detr.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            - head_inputs_dict (dict): The keyword args dictionary of the
+                bbox_head functions, which is usually empty, or includes
+                `enc_outputs_class` and `enc_outputs_class` when the detector
+                support 'two stage' or 'query selection' strategies.
+        """
+        batch_size = memory.size(1)
+        query_pos = self.query_embedding.weight
+        query_pos = query_pos.unsqueeze(1).repeat(1, batch_size, 1)
+        if self.num_patterns == 0:
+            query = torch.zeros(
```
Comment (jshilong) [inline mmdet/models/detectors/dab_detr.py:101]: query = query_pos.new_zeros(
                self.num_query,
                batch_size,
                self.embed_dims)
                )

File: mmdet/models/layers/transformer/dab_detr_transformer.py [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+from torch import Tensor
+
+from .detr_transformer import (DetrTransformerDecoder,
+                               DetrTransformerDecoderLayer,
+                               DetrTransformerEncoder,
+                               DetrTransformerEncoderLayer)
+from .utils import MLP, inverse_sigmoid
+
+
+def gen_sineembed_for_position(pos_tensor,
```
Comment (jshilong) [inline mmdet/models/layers/transformer/dab_detr_transformer.py:22]: map_position_to_sinencoding ?

File: mmdet/models/layers/transformer/dab_detr_transformer.py [hunk 6]
```diff
@@ -0,0 +1,667 @@
...
+        self.cross_attn = cross_attn
+        self.keep_query_pos = keep_query_pos
+        self.embed_dims = embed_dims
+        self.num_heads = num_heads
+        self.attn_drop = Dropout(attn_drop)
+        self.proj_drop = Dropout(proj_drop)
+
+        self._init_proj()
+
+    def _init_proj(self):
```
Comment (jshilong) [inline mmdet/models/layers/transformer/dab_detr_transformer.py:99]: just rename it to _init_layers

File: mmdet/models/layers/transformer/dab_detr_transformer.py [hunk 7]
```diff
@@ -0,0 +1,667 @@
...
+                If None, the `key` will be used.
+            attn_mask (Tensor): ByteTensor mask with shape [num_queries,
+                num_keys]. Same in `nn.MultiheadAttention.forward`.
+                Defaults to None.
+            key_padding_mask (Tensor): ByteTensor with shape [bs, num_keys].
+                Defaults to None.
+
+        Returns:
+            Tuple[Tensor]: Attention outputs of shape :math:`(L, N, E)`,
+                where :math:`L` is the target sequence length,
```
Comment (jshilong) [inline mmdet/models/layers/transformer/dab_detr_transformer.py:136]: no indentation

File: mmdet/models/layers/transformer/dab_detr_transformer.py [hunk 8]
```diff
@@ -0,0 +1,667 @@
...
+            if self.return_intermediate:
+                if self.post_norm is not None:
+                    intermediate.append(self.post_norm(output))
+                else:
+                    intermediate.append(output)
+
+        if self.post_norm is not None:
+            output = self.post_norm(output)
+
+        if reg_branches is not None and self.return_intermediate:
```
Comment (jshilong) [inline mmdet/models/layers/transformer/dab_detr_transformer.py:601]: Can we simplify the logic here

File: mmdet/models/layers/transformer/dab_detr_transformer.py [hunk 9]
```diff
@@ -0,0 +1,667 @@
...
+        Args:
+            query (Tensor): Input queries of encoder, has shape
+                (num_query, bs, dim).
+            query_pos (Tensor): The positional embeddings of the queries, has
+                shape (num_feat, bs, dim).
+            query_key_padding_mask (Tensor): ByteTensor, the key padding mask
+                of the queries, has shape (num_feat, bs).
+
+        Returns:
+            Tensor: With shape (bs, num_query, dim) if `batch_first` is `True`,
```
Comment (jshilong) [inline mmdet/models/layers/transformer/dab_detr_transformer.py:655]: Check the batch_first

Comment (jshilong) [reply mmdet/models/layers/transformer/dab_detr_transformer.py:655]: remove all batch first related comments and make batch first as the only behavior

File: mmdet/models/detectors/dab_detr.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        self.positional_encoding = SinePositionalEncoding(
+            **self.positional_encoding_cfg)
+        self.encoder = DabDetrTransformerEncoder(**self.encoder)
+        self.decoder = DabDetrTransformerDecoder(**self.decoder)
+        self.embed_dims = self.encoder.embed_dims
+        self.query_dim = self.decoder.query_dim
+        self.query_embedding = nn.Embedding(self.num_queries, self.query_dim)
+        if self.num_patterns > 0:
+            self.patterns = nn.Embedding(self.num_patterns, self.embed_dims)
+        self.nb_dec = self.decoder.num_layers
```
Comment (jshilong) [inline mmdet/models/detectors/dab_detr.py:54]: seems we do not use this attribute. I suggest removing this attribute and just using `self.decode.num_layers` at the detector level.

File: mmdet/models/layers/transformer/dab_detr_transformer.py [hunk 11]
```diff
@@ -0,0 +1,603 @@
...
+    def __init__(self,
+                 *args,
+                 query_dim=4,
+                 query_scale_type='cond_elewise',
+                 modulate_hw_attn=True,
+                 **kwargs):
+
+        self.query_dim = query_dim
+        self.query_scale_type = query_scale_type
+        self.modulate_hw_attn = modulate_hw_attn
```
Comment (jshilong) [inline mmdet/models/layers/transformer/dab_detr_transformer.py:415]: Adding `with` or `use` as a prefix may be better

**Comment by LYMDLUT** [comment]
https://github.com/open-mmlab/mmdetection/blob/f232f3d390848dff2922e8a5beb6eebe47341c92/mmdet/models/layers/transformer/dab_detr_transformer.py#L312

ref_sine_embed

**Comment by KeiChiTse** [comment]
### [Update]
Finished training mAP alignment!
**dab-detr_r50_0pat_80cls_**: _**mAP 42.3, +0.1**_ compared to reported results in paper.
**dab-detr_r50_0pat_91cls_**: _**mAP 42.7, +0.1**_ compared to reported results in paper.

**Logs** and **checkpoint** will be released in a few days.
More experiements are on the way!
More refactored detr-like algorithms are on the way!

File: mmdet/models/layers/transformer/detr_transformer.py [hunk 12]
```diff
@@ -320,18 +295,17 @@ def forward(self,
                 same shape as `key`, then `query_pos` will be used for
                 `key_pos`. Defaults to None.
             self_attn_masks (Tensor, optional): ByteTensor mask, has shape
-                (num_query, num_key), as in `nn.MultiheadAttention.forward`.
+                (num_queries, num_key), as in `nn.MultiheadAttention.forward`.
```
Comment (Li-Qingyun) [inline mmdet/models/layers/transformer/detr_transformer.py:298]: `num_keys` ?

**Review by Li-Qingyun — APPROVED** [review]

File: mmdet/models/layers/transformer/utils.py [hunk 13]
```diff
@@ -44,6 +44,58 @@ def nchw_to_nlc(x):
...
+            along x-axis or y-axis. Note the final returned dimension
+            for each position is 2 times of this value.
+        temperature (int, optional): The temperature used for scaling
+            the position embedding. Defaults to 10000.
+        scale (float, optional): A scale factor that scales the position
+            embedding. The scale will be used only when `normalize` is True.
+            Defaults to 2*pi.
+
+    Returns:
+    pos (Tensor): Returned encoded positional tensor.
```
[the remark this answers is not in the document]
Comment (KeiChiTse) [reply mmdet/models/layers/transformer/utils.py:66]: > 

The docstring of 'Returns' should be written with **no indention**.

File: mmdet/models/dense_heads/dab_detr_head.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        self.fc_cls = Linear(self.embed_dims, self.cls_out_channels)
+        # reg branch
+        self.fc_reg = MLP(self.embed_dims, self.embed_dims, 4, 3)
+
+    def init_weights(self) -> None:
+        """initialize weights."""
+        if self.loss_cls.use_sigmoid:
+            bias_init = bias_init_with_prob(0.01)
+            nn.init.constant_(self.fc_cls.bias, bias_init)
+        constant_init(self.fc_reg.layers[-1], 0., bias=0.)
```
[the remark this answers is not in the document]
Comment (KeiChiTse) [reply mmdet/models/dense_heads/dab_detr_head.py:37]: Yes, only difference is the focal loss. But I use an MLP class here to make it simpler for box refinement in decoder.

File: mmdet/models/dense_heads/dab_detr_head.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+            - layers_bbox_preds (Tensor): Sigmoid outputs from the regression
+              head with normalized coordinate format (cx, cy, w, h), has shape
+              (num_hidden_states, bs, num_query, 4).
+        """
+        layers_cls_scores = self.fc_cls(hidden_states)
+        references_before_sigmoid = inverse_sigmoid(references, eps=1e-3)
+        tmp = self.fc_reg(hidden_states)
+        tmp[..., :references_before_sigmoid.
+            size(-1)] += references_before_sigmoid
+        layers_bbox_preds = tmp.sigmoid()
```
[the remark this answers is not in the document]
Comment (KeiChiTse) [reply mmdet/models/dense_heads/dab_detr_head.py:68]: ok.

File: mmdet/models/dense_heads/dab_detr_head.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        return layers_cls_scores, layers_bbox_preds
+
+    def loss(self, hidden_states: Tensor, references: Tensor,
```
[the remark this answers is not in the document]
Comment (KeiChiTse) [reply mmdet/models/dense_heads/dab_detr_head.py:71]: okk.

**Comment by jshilong** [comment]
1. Please merge the refacor-detr after #9460 is merged， and change the data-flow to BNC
2. please remove all 91 classes related codes

File: mmdet/models/dense_heads/dab_detr_head.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Li-Qingyun) [reply mmdet/models/dense_heads/dab_detr_head.py:37]: > Yes, only difference is the focal loss. But I use an MLP class here to make it simpler for box refinement in decoder.

I mean that maybe the DETR can be modified like this.

[the remark this answers is not in the document]
Comment (KeiChiTse) [reply mmdet/models/dense_heads/dab_detr_head.py:37]: > > Yes, only difference is the focal loss. But I use an MLP class here to make it simpler for box refinement in decoder.
> 
> I mean that maybe the DETR can be modified like this.

Of course, I mentioned that during one of our meetings few months ago, denied by @jshilong then.

**Comment by KeiChiTse** [comment]
### [Update]
Double checked everything, inference and training are all set! Ready to be merged!

File: mmdet/models/dense_heads/dab_detr_head.py [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+    def _init_layers(self) -> None:
+        """Initialize layers of the transformer head."""
+        # cls branch
+        self.fc_cls = Linear(self.embed_dims, self.cls_out_channels)
+        # reg branch
+        self.fc_reg = MLP(self.embed_dims, self.embed_dims, 4, 3)
+
+    def init_weights(self) -> None:
+        """initialize weights."""
+        if self.loss_cls.use_sigmoid:
```
Comment (jshilong) [inline mmdet/models/dense_heads/dab_detr_head.py:34]: elf.loss_cls.get("use_sigmoid",True)

File: mmdet/models/layers/transformer/utils.py [hunk 18]
```diff
@@ -44,6 +46,58 @@ def nchw_to_nlc(x):
     return x.flatten(2).transpose(1, 2).contiguous()
 
 
+def convert_coordinate_to_encoding(pos_tensor: Tensor,
```
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:49]: Please help to resolve and check the comments in @LYMDLUT 's PR

File: mmdet/models/layers/transformer/utils.py [hunk 19]
```diff
@@ -385,6 +439,297 @@ def inverse_sigmoid(x: Tensor, eps: float = 1e-5) -> Tensor:
...
+                 proj_drop: float = 0.,
+                 cross_attn: bool = False,
+                 keep_query_pos: bool = False,
+                 batch_first: bool = True,
+                 init_cfg=None):
+        super().__init__(init_cfg=init_cfg)
+
+        assert batch_first is True, 'First \
+        dimension of all DETRs in mmdet is \
+        `batch`, please set `batch_first` flag.'
```
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:477]: we do not support `batch_first=False` for ConditionalAttention
@LYMDLUT  @KeiChiTse  please help to fix this

**Comment by jshilong** [comment]
Please merge the refactor-detr first and resolve all conflicts

File: mmdet/models/layers/transformer/dab_detr_transformer.py [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+                torch.stack(intermediate_reference_points),
+            ]
+        else:
+            return [
+                output.unsqueeze(0),
+                torch.stack(intermediate_reference_points)
+            ]
+
+
+class DabDetrTransformerEncoder(DetrTransformerEncoder):
```
Comment (jshilong) [inline mmdet/models/layers/transformer/dab_detr_transformer.py:263]: DAB may be better because it represents the `Dynamic Anchor Boxes `

Comment (KeiChiTse) [reply mmdet/models/layers/transformer/dab_detr_transformer.py:263]: ok 👍 :)

File: mmdet/models/layers/transformer/utils.py [hunk 21]
```diff
@@ -46,6 +46,73 @@ def nchw_to_nlc(x):
     return x.flatten(2).transpose(1, 2).contiguous()
 
 
+def convert_coordinate_to_encoding(coord_tensor: Tensor,
```
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:49]: Why do these two functions still exist in this pr? They should have been merged in Conditional DETR

File: mmdet/models/layers/transformer/utils.py [hunk 22]
```diff
@@ -371,216 +438,6 @@ def forward(self, x: Tensor,
...
-
-        pos = torch.cat((pos_y, pos_x, pos_w, pos_h), dim=-1)
-    else:
-        raise ValueError('Unknown pos_tensor shape(-1):{}'.format(
-            coord_tensor.size(-1)))
-    return pos
-
-
-@MODELS.register_module()
-class DynamicConv(BaseModule):
```
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:477]: why remove this function?

**Comment by jshilong** [comment]
LGTM

**Review by jshilong — APPROVED** [review]

File: mmdet/models/layers/transformer/utils.py (hunk 21 above)
Comment (ZwwWayne) [inline mmdet/models/layers/transformer/utils.py:49]: convert_coordinate_to_encoding -> coordinate_to_encoding

File: mmdet/models/layers/transformer/dab_detr_transformer.py [hunk 23]
```diff
@@ -0,0 +1,300 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (ZwwWayne) [inline mmdet/models/layers/transformer/dab_detr_transformer.py:1]: considering renaming the following files:
dab_detr_transformer -> dab_detr_layers.py
detr_transformer -> detr_layers.py
deformable_detr_transformer -> deformable_detr_layers.py

File: mmdet/models/dense_heads/dab_detr_head.py [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+
+from mmdet.registry import MODELS
+from mmdet.structures import SampleList
+from mmdet.utils import InstanceList
+from ..layers import MLP, inverse_sigmoid
+from .detr_head import DETRHead
+
+
+@MODELS.register_module()
+class DABDETRHead(DETRHead):
```
Comment (jshilong) [inline mmdet/models/dense_heads/dab_detr_head.py:17]: Maybe you can inherit conditionaldetrhead？

**Review by jshilong — APPROVED** [review]
LGTM

**Comment by KeiChiTse** [comment]
### [Update]
Ready to be merged...

**Pull request merged**